backend/models/upload.js

```diff
@@ -0,0 +1,301 @@
+				const size = parseInt(fileReq.headers["content-length"]);
+
+				if(isIdgn || Upload.acceptedFormat.indexOf(format.toLowerCase()) === -1) {
+					return cb({resCode: responseCodes.FILE_FORMAT_NOT_SUPPORTED });
+				}
+
+				if(size > config.uploadSizeLimit) {
+					return cb({ resCode: responseCodes.SIZE_LIMIT });
+				}
+
+				const sizeInMB = size / (1024 * 1024);
+				middlewares.freeSpace(account).then(space => {
+
+					if(sizeInMB > space) {
+						cb({ resCode: responseCodes.SIZE_LIMIT_PAY });
+					} else {
+						cb(null, true);
+					}
+				});
+			}
+		});
+
+		upload.single("file")(req, null, function (err) {
+			if (err) {
+				return reject(err);
+
+			} else if(!req.file.size) {
+				return reject(responseCodes.FILE_FORMAT_NOT_SUPPORTED);
+
+			} else {
+				modelStatusChanged(null, account, model, { status: "uploaded" });
+				return resolve(req.file);
+			}
+		});
+	});
+
+	// req.body.tag wont be defined after the file has been uploaded
+	await isValidTag(account, model, req.body.tag);
+
+	return uploadedFile;
+};
+
+Upload.initUploadChunks = async (teamspace, model, corID, username, headers) => {
+	if (!headers["x-ms-transfer-mode"] ||
+		headers["x-ms-transfer-mode"] !== "chunked" ||
+		!headers["x-ms-content-length"] ||
+		isNaN(headers["x-ms-content-length"])) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const sharedSpacePath = importQueue.getSharedSpacePath();
+	const chunkSize = Math.min(C.MS_CHUNK_BYTES_LIMIT, parseInt(headers["x-ms-content-length"]));
+
+	if (!fs.existsSync(`${sharedSpacePath}/${corID}.json`)) {
+		throw responseCodes.CORRELATION_ID_NOT_FOUND;
+	}
+
+	// upload model with tag
+	modelStatusChanged(null, teamspace, model, { status: "uploading", username });
+
+	await importQueue.mkdir(`${sharedSpacePath}/${corID}/`);
+	await importQueue.mkdir(`${sharedSpacePath}/${corID}/chunks/`);
+
+	return { "x-ms-chunk-size": chunkSize };
+};
+
+Upload.uploadChunk = async (teamspace, model, corID, req) => {
+	if (!config.cn_queue) {
+		return Promise.reject(responseCodes.QUEUE_NO_CONFIG);
+	}
+
+	if (!req.headers["content-range"]) {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const [contentRange, contentSize] = req.headers["content-range"].split("/");
+	const [sizeUnit, contentRangeValue] = contentRange.split(" ");
+
+	if (sizeUnit !== "bytes") {
+		throw responseCodes.INVALID_ARGUMENTS;
+	}
+
+	const contentMax = contentRangeValue.split("-")[1];
+
+	const sizeRemaining = contentSize - contentMax - 1;
+	const chunkSize = Math.min(C.MS_CHUNK_BYTES_LIMIT, sizeRemaining);
+
+	const sharedSpacePath = importQueue.getSharedSpacePath();
+	const timestamp = Date.now();
+
+	await handleChunkStream(req, `${sharedSpacePath}/${corID}/chunks/${timestamp}`);
+
+	systemLogger.logInfo(`CONTENT-RANGE=${req.headers["content-range"]}`);
+	systemLogger.logInfo(`CHUNKSIZE=${chunkSize}`);
+	if (chunkSize === 0) {
+		modelStatusChanged(null, teamspace, model, { status: "uploaded" });
+		await stitchChunks(corID, "upload");
```

Comment:
@carmenfan The workflow for normal uploads always wants to move a file to this `corID/filename` location. The `move` function seems to not like it if source and destination are the same and I think some more changes to distinguish between a normal upload and chunked upload would have to be made if we don't want to move the chunked file.

Reply:
I think we want to do a check and not bother with the mv if the file is already in the right place. There's not a lot of point naming something something else just because you need to do a move that doesn't need to happen